File: src/content/docs/topics/cs/leetcode-150/trees/543-diameter-of-binary-tree.py
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def build_tree(vals):
    if not vals:
        return None
    root = TreeNode(vals[0])
    q = [root]
    i = 1
    while q and i < len(vals):
        node = q.pop(0)
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            q.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            q.append(node.right)
        i += 1
    return root

def diameter_of_binary_tree(root):
    pass  # TODO: implement

def _run_tests():
    assert diameter_of_binary_tree(build_tree([1, 2, 3, 4, 5])) == 3
    assert diameter_of_binary_tree(build_tree([1, 2])) == 1
    assert diameter_of_binary_tree(build_tree([1])) == 0
    t = TreeNode(1, TreeNode(2, TreeNode(3, TreeNode(4))))
    assert diameter_of_binary_tree(t) == 3
    t2 = build_tree([1, 2, None, 3, 4])
    assert diameter_of_binary_tree(t2) == 2
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/trees/543-diameter-of-binary-tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func diameterOfBinaryTree(root *TreeNode) int {
	// TODO: implement
	return 0
}

func runTests() {
	// [1,2,3,4,5] -> 3
	root1 := &TreeNode{Val: 1,
		Left:  &TreeNode{Val: 2, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 5}},
		Right: &TreeNode{Val: 3},
	}
	assert(diameterOfBinaryTree(root1) == 3)

	// [1,2] -> 1
	root2 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}}
	assert(diameterOfBinaryTree(root2) == 1)

	// single node -> 0
	assert(diameterOfBinaryTree(&TreeNode{Val: 1}) == 0)

	// skewed tree: diameter = 3
	root4 := &TreeNode{Val: 1, Left: &TreeNode{Val: 2, Left: &TreeNode{Val: 3, Left: &TreeNode{Val: 4}}}}
	assert(diameterOfBinaryTree(root4) == 3)

	// [1,2,null,3,4] -> diameter 2 (path 3->2->4)
	root5 := &TreeNode{Val: 1,
		Left: &TreeNode{Val: 2, Left: &TreeNode{Val: 3}, Right: &TreeNode{Val: 4}},
	}
	assert(diameterOfBinaryTree(root5) == 2)

	fmt.Println("all tests pass")
}

func main() { runTests() }
